### ISSUE TITLE
[Feature Request] Full Desktop Zoom

### ISSUE THREAD

**Issue by BurhanDanger** [yshui/compton#43]
https://github.com/chjj/compton/issues/188

**Comment by rien333** [comment]
Doing this properly would be incredibly useful for visually impaired users that want to use FoSS/non-bloaty solutions. To my knowledge, no software package does desktop-zooming in a nice way (I kinda like the macOS approach), so implementing this would be of great benefit to the community as a whole and to the continuing relevance of compton.

**Comment by zezic** [comment]
> To my knowledge, no software package does desktop-zooming in a nice way

Actually, Compiz can do it almost perfectly, but it's heavy and it looks harder to maintain (so may be it is considered as a "bloaty solution"), this is why I'm also want to see this feature available in Compton.

_Should we try to create a $$ bounty for this issue? Maybe some other features are also possible to be supported by bakers, like, for example, "antialiased rounded corners"._

**Comment by nkreer** [comment]
I'd like to give this a boost once more. Having a feature like compiz' ezoom would truly enhance my experience as a visually impaired linux user. Maybe I could even implement it myself, but I'd need some pointers as to where I should start looking and how exactly something like this could be handled here.

**Comment by rien333** [comment]
> Maybe I could even implement it myself, but I'd need some pointers as to where I should start looking and how exactly something like this could be handled here.

@yshui

**Comment by YoYoYonnY** [comment]
@nkreer See #312, which should make this significantly easier (soon)

**Comment by adamhotep** [comment]
For what it's worth, the compositor for XFCE (a part of [xfwm4](https://docs.xfce.org/xfce/xfwm4/start)) can also zoom into the area under the mouse pointer using <kbd>Win</kbd>+<kbd>mousewheel</kbd>, but copying code may be tricky ([xfwm4 is GPLv2](https://gitlab.xfce.org/xfce/xfwm4/-/blob/master/COPYING) while [Compiz is MIT-licensed](https://gitlab.com/compiz/compiz-core/-/blob/master/COPYING); [picom is dual-licensed as MIT/MPL](https://github.com/yshui/picom/blob/next/COPYING)).

If somebody (@zezic?) sets up a monetary bounty for this feature, I'd be happy to contribute to it.

**Comment by Antiquis** [comment]
> For what it's worth, the compositor for XFCE (a part of [xfwm4](https://docs.xfce.org/xfce/xfwm4/start)) can also zoom into the area under the mouse pointer using Win+mousewheel, but copying code may be tricky ([xfwm4 is GPLv2](https://gitlab.xfce.org/xfce/xfwm4/-/blob/master/COPYING) while [Compiz is MIT-licensed](https://gitlab.com/compiz/compiz-core/-/blob/master/COPYING); [picom is dual-licensed as MIT/MPL](https://github.com/yshui/picom/blob/next/COPYING)).
> 
> If somebody (@zezic?) sets up a monetary bounty for this feature, I'd be happy to contribute to it.

Looks like we might FINALLY be starting to get somewhere with this! So many people have deemed full-screen desktop magnification as unnecessary / uninteresting, and it’s been an absolute nightmare for legally blind / visually impaired users with a passion for FoSS such as myself.

Let’s please get a monetary bounty setup for this! Contributing code for this is well beyond my ability, but I’d happily contribute $$ if it meant getting this feature implemented!

Thanks for sharing that update @adamhotep!

**Comment by christianparpart** [comment]
> Doing this properly would be incredibly useful for visually impaired users that want to use FoSS/non-bloaty solutions. To my knowledge, no software package does desktop-zooming in a nice way (I kinda like the macOS approach), so implementing this would be of great benefit to the community as a whole and to the continuing relevance of compton.

Speaking as a severily visually disabled user myself, that is sadly a **really** helpful feature. I cannot express that often enough.
I am currnently using KWin (which has that feature), but I'd like to use i3 (with picom) but can't switch because of that missing feature. I'd love to see here having some progress. :)

**Comment by rien333** [comment]
@christianparpart It's far from optimal, but if've bound the same set of keys (<kbd>Super</kbd> <kbd>+</kbd>, (<kbd>Super</kbd> <kbd>-</kbd>, and (<kbd>Super</kbd> <kbd>0</kbd> to reset the zoom) accross every major app that I use to perform text-scaling and zooming. I find that helps a bit (if I for whatever reason have trouble parsing my screen contents)

**Comment by yassinebridi** [comment]
For anyone who really needs this feature, i have a script that mimics this functionality.
It's not perfect, but it's usable at least.

![Peek 2021-05-19 21-03](https://user-images.githubusercontent.com/18403595/118877043-dece0f80-b8e5-11eb-950b-615e802d9e70.gif)

You can find the script here:
https://github.com/yassinebridi/.dotfiles/blob/master/bin/bin/zoom

You can just map the script using `sxhkd` for example like this:
```
super + button4
  ~/bin/zoom
```

Basically how it works, it takes a screenshot of the current window and open that screenshot in fullscreen, which then you can zoom and zoom out.

I'm using sxiv as the image viewer, because of how fast it is(which is really necessary here), and the sensible keymaps that it has.

**Comment by Antiquis** [comment]
I have started a bounty for this project:
https://app.bountysource.com/issues/65864159-feature-request-full-desktop-zoom

I'd really appreciate it if we could get some more backers, so we can get this implemented! I'll probably contribute more in a couple of days...

@zezic @rien333 @nkreer @yyny @christianparpart @adamhotep @yassinebridi

**Comment by fjfnaranjo** [comment]
I needed this (for low-vision adaptations) and ended using https://github.com/tsoding/boomer because someone recommended it in the Discord server.

If after using that project, someone thinks there is a strong reason for having this functionality natively available in picom, let me know in a comment.

**Comment by adamhotep** [comment]
@fjfnaranjo – it appears from the screenshot that boomer freezes the screen and then zooms in. I'm looking for something that lets you interact with the items still visible upon zooming (like text input or a video).

**Comment by fjfnaranjo** [comment]
> @fjfnaranjo – it appears from the screenshot that boomer freezes the screen and then zooms in. I'm looking for something that lets you interact with the items still visible upon zooming (like text input or a video).

Is exactly as you said. I started looking into composers because I wanted to use it with videogames, but for 95% of my use cases this tools works fine.

Interested in knowing what other people think about this and what do you thing about a native zoom feature in picom. Bloat or not?

**Comment by jfly** [comment]
I'd also like the ability to zoom in and still interact with items.

**Comment by jwueller** [comment]
@fjfnaranjo @jfly Looking at the README, I assume that's what the "Live image update" flag of that tool controls, so it's clearly _intended_ to be able to do that, but maybe it's broken right now.

**Comment by mxmilkiib** [comment]
aside from making it easier to share something on the screen with folk over your shoulder or in a presentation, it's a great accessibility idea for visually (or attention) impaired folk, so no not bloat, imho

**Comment by fjfnaranjo** [comment]
> ... great accessibility idea for visually (or attention) impaired folk, so no not bloat, imho

Yeah, I get that (I'm visually impaired myself). But at the same time picom default config file is already very big...

I definitely want the feature (that Pillars of Eternity II save file is getting lonely). But the option to have it in another composer, dedicated to visually impaired people, is also good.

Maybe I can try doing it and then picom can choose to merge it or not. And if they don't want to, I can just remove all the other options in the fork and have it as an accessibility alternative. Then, with time, it can evolve to have... idk... a VR render of the desktop for some headset with good resolution? Options to increase contrast? Or to apply color curves to the screen (to reduce tiredness in the eye)?

**Comment by dkaszews** [comment]
If you could be satisfied with zooming a single window instead of entire screen, I think this is already somewhat implementable using [a zoom shader](https://github.com/ikz87/picom-shaders/blob/main/Practical/zoom.glsl). I'm not sure how to conditionally activate it, the obly thing that comes to mind currently is a script that modifies `picom.conf`, but then you might have issues scrolling it.

I think the best way to handle it would be to add ability to pass custom parameters in https://github.com/yshui/picom/issues/1066, plus optionally full-screen shaders. That way you could dynamically change the zoom level and origin.

**Comment by fjfnaranjo** [comment]
Thanks! I will take this into consideration if I try to hack a solution.

**Comment by fjfnaranjo** [comment]
Well @dkaszews , your advice was put into practice and I'm back to share the initial results.

This is using last commit from "next" (https://github.com/yshui/picom/commit/c97008cd3af9fabeca2276156763fe21e12f3e58).

Config:

```
backend = "glx"
rules: ({
  match = "class_g = 'PillarsOfEternityII'";
  shader = "../../zoom.glsl";
})
```

Shader:

```glsl
#version 330

uniform sampler2D tex;

in vec2 texcoord;

float scale = 1.75;
int radius = 200;
vec2 mg_center = vec2(1200, 950);

vec4 window_shader() {
    vec4 final_color = texelFetch(tex, ivec2(texcoord), 0);
    float dist = distance(texcoord, mg_center);

    if (dist < radius) {

        // Calculate the vector from the center of the magnifying glass
        // to the current fragment.
        vec2 centered_coords = texcoord - mg_center;

        // Apply the magnification formula. This shrinks the distance from
        // the center, making the texture appear magnified.
        vec2 magnified_coords = centered_coords / scale;

        // Add the magnified coordinates back to the center of the
        // magnifying glass to get the new magnified texture coordinate.
        vec2 new_texcoord = magnified_coords + mg_center;

        // Fetch the color from the original texture at the new magnified coordinate.
        final_color = texelFetch(tex, ivec2(new_texcoord), 0);

    }

    return final_color;
}
```

This is mostly how it looks (this is a mock, in my screen you can also see the game cursor in the magnified section):

![Image](https://github.com/user-attachments/assets/0cd25b02-a9f7-4e66-83be-88907875df86)

Magic... I just have to inject the coords of the cursor and the state of one of the keys.

Being closer to play this game again makes me very happy. Thanks!

**Comment by dkaszews** [comment]
@fjfnaranjo Really happy this worked for your usecase! I think injecting variables via a script and `sed -i` should work, you might need to force picom reload e.g. with `touch picom.conf`. You probably won't get too high refresh rate and reload might cause flicker and other artifacts, but for what looks like a turn based strategy it might be enough. Best solution would probably be adding mouse coords in #1066 .

**Comment by fjfnaranjo** [comment]
Yes! For now I'm just going to investigate how can I get the cursor/keyboard state from the X11 lib. If I can get them, I will just add them as input parameters to the shaders and I'll fork and follow upstream.

I don't see much movement on the implementation of #1066 and I don't think my usecase is by itself sufficient to justify adding parameters to the shader. Issue/PRs requesting more parameters can easily lead to scope creep. And I think picom "feels" more like a ricing tool than like an accessibility tool.

I think I would really like to see #312 becoming a fork. Dropping xrender and moving all the effects to shaders looks reasonable. That way, my usecase will just become a module, instead of creeping up the scope. And the new project will become even lighter and easier to maintain.

**Comment by schrmh** [comment]
@fjfnaranjo If you do not need fancy stuff aside maybe transparency within a compositor then it should be possible to get a compositor within about 1k lines of C code – according to cloc.
Here is one that implements a magnifier similar to [compiz' simple (rectangular) magnifier](https://gitlab.com/compiz/compiz-plugins-main/-/raw/master/src/mag/mag.c) that magnifies an area around the mouse pointer – you could increase MAG_SIZE to get something similar to full desktop zoom but with a too big size it will lag.

Used a bit of my knowledge and fed Claude parts from Xlib documentation as well as [QHotKey](https://raw.githubusercontent.com/Skycoder42/QHotkey/refs/heads/master/LICENSE) and [fastcompmgr](https://raw.githubusercontent.com/tycho-kirchner/fastcompmgr/refs/heads/master/LICENSE) and xcompmgr code until I finally got a non-flickering magnifier that has no staircase effect and that works everywhere AND that allows me to still click to windows. (Years ago I never was able to get all of this at the same time, especially the staircase problem was annoying and I always got glitches somehow, and recently my concentration and other irl issues got way too worse and I want to do & have way too much stuff working the way I want so attempting this completely on my own without Claude was out of the question for now.)

[compositor.c.txt](https://github.com/user-attachments/files/21822470/compositor.c.txt)
(since code was making this comment so long, I decided to just replace it with a file ~~of the same content~~. Also... .c extension is not supported on GitHub?!)


Oh also bugs with this code:
~- **Context menu rendering issues in Firefox-based browsers and on window decorations (at least on FVWM3)** (first of that I just noticed after 40 min of sending this comment. Will try to fix this later)~
- After scrolling up and down quickly in a window and then trying to magnify something, content will still slightly get scrolled up.
- When x1.0 and on the side, it can leave a picture when deactivating zoom.
Latter can likely pretty easily be fixed.

Will probably push this to codeberg at some point when it is more modular cause I want a lot of stuff in a compositor that I have either seen nowhere or on other platforms but never on X.

[0001-Add-magnifier.patch.txt](https://github.com/user-attachments/files/21821869/0001-Add-magnifier.patch.txt)
**(Due to the bold-marked bug with context menus, here a patch for xcompmgr to add magnification there. Added stuff pretty much together instead of following the style but w/e. Also I feel like window movement is slower with this added compared to above code; something can probably be optimized)**

**Comment by fjfnaranjo** [comment]
Thanks @schrmh . This is also a good example to understand how a compositor can work if someone, as me, doesn't have experience in this kind of projects.

I like the single file, C, 0-depends, suckless style ❤️ 

I was thinking of keeping glx instead of XRender because I intent to use this for videogames and I've heard of XRender having a severe impact on performance/input lag. But the reality is that I didn't try to verify this claims.

I will test this program with demanding games and see how they perform.

**Comment by schrmh** [comment]
Oh, main reason this – likely not so good – code uses XRender is that I never came across any magnification code that uses it. And I find it a bit simpler than OpenGL or Vulkan. But yes, I think the latter two are to be preferred.
~(Also see my edit with the additional bug I noticed that is a bit more annoying than the two others; that's why I made it bold formatted.)~

**Comment by fjfnaranjo** [comment]
I need to run a proper test, checking at system metrics, but my CPU was able to handle this zoom effect without a big impact in performance for a couple of games. It got quite hot thou, so, not the safest for videogames.

With a couple of edits I was able to quickly adapt it to my zoom preferences.

Another thing I noticed, is that I have lots of things capturing events from X11. I will probably have to check how to make this work with fullscreen games, because in my tests I had to get out of the game to toggle and adjust the zoom.

**Comment by yshui** [comment]
I'd like to hear how people are using zooming. i'd imagine you would want to zoom an area around your cursor? or do you want to zoom a fixed area on your screen?

**Comment by schrmh** [comment]
I think that kind of depends on the person and their use cases.
I think what I have seen the most in X11 desktop environments is zoom of whole desktop and per monitor. Personally I prefer the around cursor thing due to least amount of screen space used – can also be useful while presenting to make it easier for the (at worst phone-using) viewer to see what the presenter is looking at currently.
When reading a long text it might be useful to zoom a monitor instead so that there is not a lot of mouse movement necessary. I can imagine some people might want to have the option to then additionally zoom some small part around the cursor but I have seen no implementation of that.
And yeah, some people might just prefer the fixed, docked to an area thing. Iirc the OS Windows does ship with three magnification modes so I guess it would be good to eventually offer at least those three.

Well, for me personally I do not really need glasses, just sitting a bit away from screens, eyes are tired sometimes and occasionally I want too see something that is small a bit better for a moment... that might also be content of a program that just does not scale well.
One of my impaired – and one-side blind – family members uses a big physical magnifier instead of a magnification mode on the OS Windows and told me especially that it annoys them when stuff is not on screen. Even when they saw way worse for months after an operation. However, they don't do much on PC, mainly using two sites for entertainment afaik. One other family member with very strong glasses  just has font size and resolution adjusted to make everything big on Windows at work and is additionally zooming in with pinch on their Android phone when reading.

**Comment by fjfnaranjo** [comment]
> I'd like to hear how people are using zooming. i'd imagine you would want to zoom an area around your cursor? or do you want to zoom a fixed area on your screen?

On the working laptop the only option I have (and this is usually good for websites with difficulty handling zoom or apps without text big enough) is a fixed full screen zoom. You press Cmd (or Ctrl...) and then you use the mouse wheel to do zoom and the borders of the screen area to pan as if you were in a RTS game (zoom doesn't follow the cursor).

I also have other potential usages of zoom I want to try specific for videogames, like following the cursor or allowing independent zoom on sections of the screen. I'm still investigating this and I'm not sure if I have a final design idea for it.
